### DISCUSSION TITLE
BUG: fail to infer other getters type in getter

### DISCUSSION THREAD

**Discussion by LoveofRedMoon** [vuejs/pinia#1076] [Help and Questions] [answered]
reproduction url: https://stackblitz.com/edit/github-ev9fpm?file=src/stores/counter.ts

In both two ways, the getter `menuRights` cannot infer getter `rights`
wish to be `UserInfo['rights']`
real be `(state) => UserInfo['rights']`

**Comment by posva** [comment, marked as answer]
As documented, you need to type the return when using `this.` in getters;

```js
getters: {
    rights: (state) => state.userInfo?.rights,
    menuRights(): UserInfo['rights'] | undefined {
      return this.rights?.filter((v) => v.type === 0);
    },
  },
```

**Reply by LoveofRedMoon** [reply]
As the code you mentioned. getter `rights` did not use `this.`, so, there is no need to add returnType to `rights`

But when `rights` used in `memuRights` getter function, ts infer `this.rights` be `() => UserInfo['rights'] | undefined`

It does nothing to do with `menuRights` returnType

**Reply by andreyan-andreev** [reply]
Does not work for me. Had to force type it as a workaround.

```typescript 
getters: {
    rights: (state) => state.userInfo?.rights,
    menuRights(): UserInfo['rights'] | undefined {

      const rights = this.rights as any[]; 
  // or const rights = <any[]>this.rights;

      return rights.filter((v) => v.type === 0);
    },
  },
```

**Reply by the-mwolf** [reply]
It works for me, but only if I add return types to all getters in my store. If just one return type is missing, no getter types will be found.

**Comment by 209** [comment]
Hello, @posva! Pinia is gorgeous! 

Could you check similar case, pls?
https://stackblitz.com/edit/github-ev9fpm-feekun?file=src%2Fstores%2Fcounter.ts
<img width="616" alt="Снимок экрана 2022-08-06 в 04 58 49" src="https://user-images.githubusercontent.com/4629854/183229195-7279af6a-64c4-45d0-b45d-34ca35395de4.png">

And without dynamic:

<img width="544" alt="Снимок экрана 2022-08-06 в 04 48 22" src="https://user-images.githubusercontent.com/4629854/183229225-f75cf33e-1735-4d99-8396-f3e264447276.png">

**Reply by vis97c** [reply]
Same thing happening to me

**Reply by BobidaHombre** [reply]
Got the same error :(((

**Reply by AxelConceicao** [reply]
same here :c

**Reply by 209** [reply]
Guys, some of the problems will go away if you update the dependencies.
But not all.

**Comment by rocifier** [comment]
what do I tell my colleagues about this in our code review? and also how can I access an action from a getter?

**Reply by posva** [reply]
you can't access an action in a getter. They shouldn't have side effects, make your getter an action so it can use other actions
